### DISCUSSION TITLE
How I can subscribe to collection in +page.server.ts in svltekit

### DISCUSSION THREAD

**Discussion by islamzaoui** [pocketbase/pocketbase#2909] [Q&A]
hello everyone
I want to subscribe to collection from the server side (Load function) in my app because I stored pb in locals
is there any way to do that?
this is my load function in my web app:
```
export const load = (async (event) => {
    await event.parent()
    if (!event.locals.pb?.authStore.isValid) {
        throw redirect(303, '/')
    }
    let messages: any[] = []
    try {
        const list = await event.locals.pb?.collection('messages').getList(1, 50, {
            sort: 'created',
            expand: 'user'

        })
        messages = list.items
    }
    catch (error) {
        messages = []
    }
    const form = await superValidate(event, msgschema)
    return { form, messages: messages.map((msg) => msg.export()) }
}) satisfies PageServerLoad;
```

**Comment by ganigeorgiev** [comment]
By "_subscribe to collection_" you mean the realtime subscriptions, right? If that's the case then I don't think you can do that in a server-side load function because the realtime subscription requires a persistent/long running connection and I'm not sure how the SvelteKit `load` function will even resolve in this case, aka. wouldn't that require blocking the user response?

I guess, you can use the realtime subscriptions in a SvelteKit server-side but you'll need:
1. `EventSource` polyfill (_if not already available_):
   ```js
   // for server: npm install eventsource --save
   import eventsource from 'eventsource';
   
   global.EventSource = eventsource;
   ```
   
2. To implement a SvelteKit server-side route/action/etc. that will serve as a proxy, aka. your users will connect to your SvelteKit node route, and inside that route you'll connect to PocketBase.


Keep in mind that when using SvelteKit with SSR you no longer have "just frontend". SvelteKit is a full-blown fullstack framework on its own and you have to decide how do you want your users to interact with PocketBase (or any other external backend service):

- Only client-side in the browser, aka. the users send direct requests to PocketBase.
- Only server-side, aka. the users send requests to your node application and the node application proxy/forward the request with some extra validation/modifications/etc. to PocketBase (_this is similar to the example shown in the [SSR sample guide](https://github.com/pocketbase/js-sdk#ssr-integration)_).
- Mixed, aka. both browser and server. In this case you'll need to maintain 2 different types of PocketBase JS SDK instances - one in the browser (could be global) and one in node (unique per request). The auths state between those 2 instances can be synced using a cookie (it will have to be exported with `httpOnly: false` and from the browser side will have to be set using something like `document.cookie`).

**Reply by islamzaoui** [reply]
Thx for help, i will try this

**Comment by realJogicodes** [comment]
I am trying to make a client-side subscription in a svelte kit app happen and am failing so far. 

Is there a trick to this? I am trying something similar to this post handling auth on the server and returning the cookie in the header and then on my client, i instantiate pb again and that  always authenticates as guest even though I *seem* to be logged in (cookie in place, front end shows protected stuff). But on the client, pb instances seem to not parse the cookie that server side authentication returns. I tried  forcing authRefresh because of it on the front end, for which the api returns 401 Unauthorized

Are there conflicts between client and server instances?

```
Uncaught (in promise) ClientResponseError 401: The request requires valid record authorization token to be set.
    _ClientResponseError ClientResponseError.ts:13
    send Client.ts:365
    promise callback*send Client.ts:363
    authRefresh RecordService.ts:578
    getUsers +page.svelte:9
    listen dom.js:361
    listen_dev dev.js:133
    mount +page.svelte:67
    mount +page.svelte:218
    m svelte-hooks.js:291
    mount_component Component.js:44
    mount root.svelte:272
    mount +layout.svelte:34
    m svelte-hooks.js:291
    mount_component Component.js:44
    mount root.svelte:171
    mount root.svelte:479
    mount_component Component.js:44
    init Component.js:162
    Root root.svelte:708
    createProxiedComponent svelte-hooks.js:341
    ProxyComponent proxy.js:242
    Proxy<Root> proxy.js:349
    initialize client.js:402
    navigate client.js:1345
    _goto client.js:346
    goto client.js:1667
    start client.js:277
    <anonymous> (index):24
    promise callback* (index):23
[ClientResponseError.ts:13:8](http://192.168.86.36:5173/node_modules/.pnpm/pocketbase@0.20.2/node_modules/pocketbase/src/ClientResponseError.ts)
```

**Reply by ganigeorgiev** [reply]
@proofofjogi It is difficult to pinpoint the exact cause of the issue without a code sample how you are initializing the client-side SDK instance.

With that said you can check the other discussions related to mixed client-side and server-side SDK access for example in https://github.com/pocketbase/pocketbase/discussions/3462, https://github.com/pocketbase/pocketbase/discussions/3837, https://github.com/pocketbase/pocketbase/discussions/3792, https://github.com/pocketbase/pocketbase/discussions/2046, https://github.com/pocketbase/pocketbase/discussions/926 and probably others.

It is really tiredsome to read the same question over and over especially when I really don't know how to explain it better. I asked several times why users even prefer to use JavaScript SSR and the only sort of an answer I got was that it "feels more safe" which to me doesn't make sense but I don't want to repeat myself (I think the most recent discussion related to this was in https://github.com/pocketbase/pocketbase/discussions/3792).

**Reply by realJogicodes** [reply]
Thanks, that was a great hint. I appreciate what you do and your answers. 
The reason why I used SSR so far in the application is that it came like this by default and I failed to consider the bigger picture as I was blindly following a svelte kit  with pocketbase tutorial. That explained how to set everything up with SSR, they did not use subscriptions though, which need to run client side. I don't have a need for SSR in what I need out of the app and will just try to move all of the pb code to the client side. I will re-architect this and try again, thank you very much.

**Reply by jonshipman** [reply]
I have this boilerplate https://github.com/jonshipman/sveltekit-pb-boilerplate
Does server SSR and client CSR using the cookie sharing method @ganigeorgiev mentioned. 

My preference for JavaScript SSR stems from the ability to report accurate HTTP responses for different queries. E.g. 404 when an entity is not found or 403 when an entity is forbidden. Yes you can do the UX/user side purely in CSR, but if you want/need proper HTTP responses then you need to do things on the server. If you also need access to realtime data, just share the cookie from server to client. Realtime data is the only reason (IMO) to query pocketbase on the client. When I don't need realtime data, I'll often just set httpOnly to true and interface with Pocketbase behind a proxy.
